File: products/LoanHub/apps/frontend/api/institutionGovernance.ts
import { api } from "@/lib/api";
import type { InstitutionType } from "@/store/slices/companiesSlice";
import type { UserRole } from "@/types/auth";

export type InstitutionGovernanceProfile = {
    id: string;
    company_id: string;
    regulator_name: string | null;
    regulatory_license_category: string | null;
    license_expiry_date: string | null;
    bank_code: string | null;
    swift_bic: string | null;
    aml_cft_officer_name: string | null;
    aml_cft_officer_email: string | null;
    data_protection_officer_name: string | null;
    data_protection_officer_email: string | null;
    regulatory_reporting_contact_email: string | null;
    complaints_contact: string | null;
    privacy_notice_url: string | null;
    data_retention_months: number;
    consent_management_enabled: boolean;
    data_export_enabled: boolean;
    ai_decisioning_enabled: boolean;
    ai_human_review_required: boolean;
    ai_explainability_required: boolean;
    ai_bias_monitoring_enabled: boolean;
    govstack_interoperability_status: "not_started" | "planned" | "in_progress" | "ready";
    dpg_readiness_status: "not_started" | "assessment" | "remediation" | "ready_for_review";
    open_api_published: boolean;
    low_bandwidth_supported: boolean;
    accessibility_reviewed: boolean;
    english_sesotho_supported: boolean;
    business_continuity_tested: boolean;
    incident_response_tested: boolean;
    interoperability_notes: string | null;
    created_at: string;
    updated_at: string;
};

export type InstitutionGovernanceUpdate = Partial<Omit<
    InstitutionGovernanceProfile,
    "id" | "company_id" | "created_at" | "updated_at"
>>;

export type ReadinessControl = {
    key: string;
    category: string;
    label: string;
    state: "complete" | "attention" | "not_applicable";
    required: boolean;
    detail: string;
};

export type InstitutionReadiness = {
    company_id: string;
    institution_type: InstitutionType;
    score_percent: number;
    completed_required_controls: number;
    required_controls: number;
    controls: ReadinessControl[];
    assigned_roles: UserRole[];
    missing_recommended_roles: UserRole[];
};

export type InstitutionRole = {
    role: UserRole;
    label: string;
    scope: "institution" | "branch" | "institution_or_branch";
    purpose: string;
};

export type GovStackCapability = {
    key: string;
    label: string;
    status: "implemented" | "integration_ready" | "planned";
    implementation: string;
};

function oversightConfig(companyId?: string) {
    return companyId ? { params: { company_id: companyId } } : undefined;
}

export async function getInstitutionGovernanceProfile(
    companyId?: string,
): Promise<InstitutionGovernanceProfile> {
    const response = await api.get<InstitutionGovernanceProfile>(
        "/institution-governance/profile",
        oversightConfig(companyId),
    );
    return response.data;
}

export async function updateInstitutionGovernanceProfile(
    payload: InstitutionGovernanceUpdate,
    companyId?: string,
): Promise<InstitutionGovernanceProfile> {
    const response = await api.put<InstitutionGovernanceProfile>(
        "/institution-governance/profile",
        payload,
        oversightConfig(companyId),
    );
    return response.data;
}

export async function getInstitutionReadiness(
    companyId?: string,
): Promise<InstitutionReadiness> {
    const response = await api.get<InstitutionReadiness>(
        "/institution-governance/readiness",
        oversightConfig(companyId),
    );
    return response.data;
}

export async function getInstitutionRoles(): Promise<InstitutionRole[]> {
    const response = await api.get<InstitutionRole[]>("/institution-governance/roles");
    return response.data;
}

export async function getGovStackCapabilities(): Promise<GovStackCapability[]> {
    const response = await api.get<GovStackCapability[]>(
        "/institution-governance/govstack-capabilities",
    );
    return response.data;
}

File: products/LoanHub/apps/backend/database/schemas/institution_governance.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from database.models.enums import InstitutionType, UserRole


ReadinessState = Literal["complete", "attention", "not_applicable"]


class InstitutionGovernanceProfileUpdate(BaseModel):
    regulator_name: str | None = Field(default=None, max_length=200)
    regulatory_license_category: str | None = Field(default=None, max_length=160)
    license_expiry_date: date | None = None
    bank_code: str | None = Field(default=None, max_length=40)
    swift_bic: str | None = Field(default=None, max_length=20)

    aml_cft_officer_name: str | None = Field(default=None, max_length=200)
    aml_cft_officer_email: EmailStr | None = None
    data_protection_officer_name: str | None = Field(default=None, max_length=200)
    data_protection_officer_email: EmailStr | None = None
    regulatory_reporting_contact_email: EmailStr | None = None
    complaints_contact: str | None = Field(default=None, max_length=200)

    privacy_notice_url: str | None = Field(default=None, max_length=500)
    data_retention_months: int | None = Field(default=None, ge=1, le=120)
    consent_management_enabled: bool | None = None
    data_export_enabled: bool | None = None

    ai_decisioning_enabled: bool | None = None
    ai_human_review_required: bool | None = None
    ai_explainability_required: bool | None = None
    ai_bias_monitoring_enabled: bool | None = None

    govstack_interoperability_status: Literal[
        "not_started", "planned", "in_progress", "ready"
    ] | None = None
    dpg_readiness_status: Literal[
        "not_started", "assessment", "remediation", "ready_for_review"
    ] | None = None
    open_api_published: bool | None = None
    low_bandwidth_supported: bool | None = None
    accessibility_reviewed: bool | None = None
    english_sesotho_supported: bool | None = None
    business_continuity_tested: bool | None = None
    incident_response_tested: bool | None = None
    interoperability_notes: str | None = Field(default=None, max_length=5000)


class InstitutionGovernanceProfileRead(InstitutionGovernanceProfileUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    created_at: datetime
    updated_at: datetime


class InstitutionRoleRead(BaseModel):
    role: UserRole
    label: str
    scope: Literal["institution", "branch", "institution_or_branch"]
    purpose: str


class ReadinessControlRead(BaseModel):
    key: str
    category: str
    label: str
    state: ReadinessState
    required: bool
    detail: str


class InstitutionReadinessRead(BaseModel):
    company_id: UUID
    institution_type: InstitutionType
    score_percent: int
    completed_required_controls: int
    required_controls: int
    controls: list[ReadinessControlRead]
    assigned_roles: list[UserRole]
    missing_recommended_roles: list[UserRole]


class GovStackCapabilityRead(BaseModel):
    key: str
    label: str
    status: Literal["implemented", "integration_ready", "planned"]
    implementation: str
